r1cs-std/src/bits/uint64.rs

```diff
@@ -0,0 +1,544 @@
+    pub fn to_bits_le(&self) -> Vec<Boolean> {
+        self.bits.clone()
+    }
+
+    /// Converts a little-endian byte order representation of bits into a
+    /// `UInt64`.
+    pub fn from_bits_le(bits: &[Boolean]) -> Self {
+        assert_eq!(bits.len(), 64);
+
+        let bits = bits.to_vec();
+
+        let mut value = Some(0u64);
+        for b in bits.iter().rev() {
+            value.as_mut().map(|v| *v <<= 1);
+
+            match b {
+                &Boolean::Constant(b) => {
+                    if b {
+                        value.as_mut().map(|v| *v |= 1);
+                    }
+                },
+                &Boolean::Is(ref b) => match b.get_value() {
+                    Some(true) => {
+                        value.as_mut().map(|v| *v |= 1);
+                    },
+                    Some(false) => {},
+                    None => value = None,
+                },
+                &Boolean::Not(ref b) => match b.get_value() {
+                    Some(false) => {
+                        value.as_mut().map(|v| *v |= 1);
+                    },
+                    Some(true) => {},
+                    None => value = None,
+                },
+            }
+        }
+
+        Self { value, bits }
+    }
+
+    pub fn rotr(&self, by: usize) -> Self {
+        let by = by % 64;
+
+        let new_bits = self
+            .bits
+            .iter()
+            .skip(by)
+            .chain(self.bits.iter())
+            .take(64)
+            .cloned()
+            .collect();
+
+        UInt64 {
+            bits:  new_bits,
+            value: self.value.map(|v| v.rotate_right(by as u32)),
+        }
+    }
+
+    /// XOR this `UInt64` with another `UInt64`
+    pub fn xor<ConstraintF, CS>(&self, mut cs: CS, other: &Self) -> Result<Self, SynthesisError>
+    where
+        ConstraintF: Field,
+        CS: ConstraintSystem<ConstraintF>,
+    {
+        let new_value = match (self.value, other.value) {
+            (Some(a), Some(b)) => Some(a ^ b),
+            _ => None,
+        };
+
+        let bits = self
+            .bits
+            .iter()
+            .zip(other.bits.iter())
+            .enumerate()
+            .map(|(i, (a, b))| Boolean::xor(cs.ns(|| format!("xor of bit_gadget {}", i)), a, b))
+            .collect::<Result<_, _>>()?;
+
+        Ok(UInt64 {
+            bits,
+            value: new_value,
+        })
+    }
+
+    /// Perform modular addition of several `UInt64` objects.
+    pub fn addmany<ConstraintF, CS>(mut cs: CS, operands: &[Self]) -> Result<Self, SynthesisError>
+    where
+        ConstraintF: PrimeField,
+        CS: ConstraintSystem<ConstraintF>,
+    {
+        // Make some arbitrary bounds for ourselves to avoid overflows
+        // in the scalar field
+        assert!(ConstraintF::Params::MODULUS_BITS >= 128);
+        assert!(operands.len() >= 2); // Weird trivial cases that should never happen
```

Comment:
Perhaps if operands.len() == 1, make it return operands[0]? 

Otherwise thats an edge-case the caller may have to handle.